File: snake-ai/GaModel.ts
import { utils } from "snake-game/utils";
import InputLayer from "./InputLayer";
import { CalcUtils } from "./CalcUtils";
import SnakeBrain from "./SnakeBrain";
import { generateLayerShape } from "./generateLayerShape";
import MultiThreadGames from "./MultiThreadGames";
import type { ActivationFunction } from "./CalcUtils";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const tempPrint: any = [];

export interface Options {
  worldWidth: number;
  worldHeight: number;
  snakeBrainConfig: {
    hiddenLayersLength: number[];
    hiddenLayerActivationFunction: ActivationFunction;
  };
  gaConfig: {
    populationSize: number;
    surviveRate: number;
    mutationRate: number;
    mutationAmount: number;
    trialTimes: number;
  };
}

interface IPopulation {
  snakeBrain: SnakeBrain;
  snakeLength: number;
  moves: number;
  fitness: number;
  survive: boolean;
  gameRecords: string;
}

type Population = IPopulation[];

export default class GaModel {
  public static fitness(moves: number, snakeLength: number, maxPossibleSnakeLength = 400): number {
    if (snakeLength === 0) return 0;

    const movesPerFood = moves / snakeLength;
    const ratioOfLength = snakeLength / maxPossibleSnakeLength;

    // for short snake
    const moveScore = moves * 10;
    const cyclicPenalty = moves * 6 * (1 - snakeLength / maxPossibleSnakeLength) ** 2;

    // for long snake, encourage fast eat food
    const lengthBaseScore = (snakeLength - 1) ** (2 + 6 * ratioOfLength) * maxPossibleSnakeLength;
    const lengthScore1 = lengthBaseScore * movesPerFood;

    // for long snake with survival more important
    const lengthScore2 = lengthBaseScore * moves * (1 - snakeLength / maxPossibleSnakeLength) ** 2;

    return moveScore - cyclicPenalty + lengthScore1 + lengthScore2;
  }

  public static spinRouletteWheel(options: IPopulation[]): IPopulation {
    const totalScore = options.reduce((acc, option) => acc + option.fitness, 0);
    let randomNum = Math.random() * totalScore;
    for (let i = 0; i < options.length; i++) {
      if (randomNum < options[i].fitness) {
        return options[i];
      }
      randomNum -= options[i].fitness;
    }
    return options[options.length - 1];
  }

  public readonly worldWidth: number;
  public readonly worldHeight: number;
  public readonly hiddenLayersLength: number[];
  public readonly hiddenLayerActivationFunction: ActivationFunction;
  public readonly populationSize: number;
  public readonly surviveRate: number;
  public readonly mutationRate: number;
  public readonly mutationAmount: number;
  public readonly trialTimes: number;

  private _generation: number;
  private readonly _numberOfSurvival: number;

  private readonly population: Population;

  private readonly multiThreadGames: MultiThreadGames;

  constructor(option: Options) {
    this._generation = -1;

    this.worldHeight = option.worldHeight;
    this.worldWidth = option.worldWidth;
    this.hiddenLayersLength = option.snakeBrainConfig.hiddenLayersLength;
    this.hiddenLayerActivationFunction = option.snakeBrainConfig.hiddenLayerActivationFunction;
    this.populationSize = option.gaConfig.populationSize;
    this.surviveRate = option.gaConfig.surviveRate;
    this.mutationRate = option.gaConfig.mutationRate;
    this.mutationAmount = option.gaConfig.mutationAmount;
    this.trialTimes = option.gaConfig.trialTimes;
    this._numberOfSurvival = Math.floor(this.populationSize * this.surviveRate);

    if (this._numberOfSurvival < 2) throw Error("Survival less than 2, please increase survive rate or population size.");

    const inputLayerLength = InputLayer.inputLayerLength;
    const layerShapes = generateLayerShape(inputLayerLength, ...this.hiddenLayersLength, SnakeBrain.OUTPUT_LAYER_LENGTH);

    this.population = Array(this.populationSize)
      .fill(null)
      .map(() => ({
        snakeBrain: new SnakeBrain({
          inputLength: inputLayerLength,
          layerShapes,
          hiddenLayerActivationFunction: this.hiddenLayerActivationFunction,
        }),
        fitness: 0,
        snakeLength: 1,
        moves: 0,
        survive: true,
        gameRecords: "",
      }));

    this.multiThreadGames = new MultiThreadGames();
  }

  public get generation() {
    return this._generation;
  }

  public get numberOfSurvival(): number {
    return this._numberOfSurvival;
  }

  public get basicConfig() {
    return {
      generation: this.generation,
      worldHeight: this.worldHeight,
      worldWidth: this.worldWidth,
      hiddenLayersLength: this.hiddenLayersLength,
      hiddenLayerActivationFunction: this.hiddenLayerActivationFunction,
      populationSize: this.populationSize,
      surviveRate: this.surviveRate,
      mutationRate: this.mutationRate,
      mutationAmount: this.mutationAmount,
      trialTimes: this.trialTimes,
      _numberOfSurvival: this._numberOfSurvival,
    };
  }

  public exportModal() {
    return JSON.stringify(this);
  }

  public async evolve(exportModal: boolean): Promise<{
    generation: number;
    population: Population;
    finalBestPlayer: IPopulation;
    bestSnakeBrain: SnakeBrain;
    bestGame: string;
    gaPlayer?: string;
  }> {
    const tempTime = new Date().getTime();

    await this.evaluate();
    this.select();
    this.crossover();
    this.mutate();

    const generation = this.generation;
    const finalBestPlayer = this.population[0];
    const bestSnakeBrain = finalBestPlayer.snakeBrain;
    const bestGame = finalBestPlayer.gameRecords;
    const population = this.population;

    const print = {
      generation,
      highestFitness: finalBestPlayer.fitness,
      meanFitness: population.reduce((acc, cur) => acc + cur.fitness, 0) / population.length,
      bestSnakeLength: finalBestPlayer.snakeLength,
      meanSnakeLength: population.reduce((acc, cur) => acc + cur.snakeLength, 0) / population.length,
      bestMoves: finalBestPlayer.moves,
      meanMoves: population.reduce((acc, cur) => acc + cur.moves, 0) / population.length,
      time: (new Date().getTime() - tempTime) / 1000,
    };

    tempPrint.push(print);
    console.table(tempPrint);

    if (exportModal) {
      return {
        generation,
        population,
        finalBestPlayer,
        bestSnakeBrain,
        bestGame,
        gaPlayer: this.exportModal(),
      };
    } else {
      return { generation, population, finalBestPlayer, bestSnakeBrain, bestGame };
    }
  }

  private async evaluate(): Promise<void> {
    this._generation++;

    const promise = this.population.map((p) => this.multiThreadGames.playGames(this.worldWidth, this.worldHeight, this.trialTimes, p.snakeBrain));
    const result = await Promise.all(promise);

    for (let i = 0; i < this.population.length; i++) {
      const p = this.population[i];
      const { snakeLengthArr, movesArr, gameRecordArr } = result[i];

      const fitnessArr: number[] = [];
      for (let j = 0; j < this.trialTimes; j++) {
        fitnessArr.push(GaModel.fitness(movesArr[j], snakeLengthArr[j]));
      }

      p.snakeLength = CalcUtils.stats.meanOfArray(snakeLengthArr);
      p.moves = CalcUtils.stats.meanOfArray(movesArr);
      p.fitness = CalcUtils.stats.meanOfArray(fitnessArr);
      p.gameRecords = gameRecordArr[CalcUtils.indexOfMaxValueInArray(fitnessArr)];
    }
  }

  private select() {
    // sort descending (highest fitness first)
    this.population.sort((a, b) => b.fitness - a.fitness);
    this.population.forEach((p, index) => {
      p.survive = index < this._numberOfSurvival;
    });
  }

  private crossover() {
    this.population.forEach((p, childIdx) => {
      if (this.population[childIdx].survive) return;

      const parent1 = this.pickParentIdx(childIdx);
      const parent2 = this.pickParentIdx(childIdx, parent1);
      p.snakeBrain.crossOver(parent1.snakeBrain, parent2.snakeBrain);
    });
  }

  private pickParentIdx(childIdx: number, anotherParent?: IPopulation): IPopulation {
    const anotherParentIdx = anotherParent ? this.population.findIndex((p) => p === anotherParent) : -1;
    const filteredPopulation = this.population.filter((p, idx) => idx !== childIdx && idx !== anotherParentIdx);
    return GaModel.spinRouletteWheel(filteredPopulation);
  }

  private mutate() {
    this.population.forEach((p, index) => {
      // prevent the best 5% from mutating
      if (index < this._numberOfSurvival * 0.05) return;

      // todo the first rate should be separate from the second rate for fine control
      if (utils.randomBool(this.mutationRate)) {
        p.snakeBrain.mutate(this.mutationRate, this.mutationAmount);
      }
    });
  }
}

File: kotlin-root/gasnake/src/main/kotlin/com/example/snake/ai/ga/GaModel.kt
package com.example.snake.ai.ga

import com.example.snake.ai.ActivationFunction
import com.example.snake.ai.InputLayer
import com.example.snake.ai.ProvidedWeightsAndBiases
import com.example.snake.ai.SnakeBrain
import com.example.snake.ai.multithread.MultiThreadGames
import com.example.snake.game.SnakeGame
import com.example.snake.utils.CalcUtils
import kotlin.math.pow
import kotlin.random.Random

class GaModel(options: Options) {
    companion object {
        fun generateLayerShape(vararg args: Int): List<List<Int>> = List(args.size - 1) { index ->
            listOf(args[index + 1], args[index])
        }

        fun fitness(moves: Int, snakeLength: Int, maxPossibleSnakeLength: Int): Double {
            if (snakeLength == 1) return 0.0

            val ratioOfLength = snakeLength.toDouble() / maxPossibleSnakeLength

            // for short snake
            val moveScore = moves * 10
            val cyclicPenalty = moves * 6 * (1 - ratioOfLength).pow(2)

            // for long snake
            val lengthScore = (snakeLength - 1).toDouble().pow(2 + 6 * ratioOfLength) * maxPossibleSnakeLength * moves

            return moveScore - cyclicPenalty + lengthScore
        }

        fun calculateCumulativeSum(values: List<Double>): List<Double> {
            val prefixSum = mutableListOf<Double>()
            var sum = 0.0
            for (x in values) {
                sum += x
                prefixSum.add(sum)
            }
            return prefixSum
        }

        fun spinRouletteWheel(candidates: List<Individual>, prefixSum: List<Double>): Individual {
            val totalScore = prefixSum.last()
            val randomNum = Random.nextDouble() * totalScore
            val index = CalcUtils.binarySearch(prefixSum, randomNum)
            return candidates[index]
        }
    }

    val worldWidth: Int = options.worldWidth
    val worldHeight: Int = options.worldHeight
    val hiddenLayersLength: List<Int> = options.snakeBrainConfig.hiddenLayersLength
    val hiddenLayerActivationFunction: ActivationFunction = options.snakeBrainConfig.hiddenLayerActivationFunction
    val populationSize: Int = options.gaConfig.populationSize
    val surviveRate: Double = options.gaConfig.surviveRate
    val populationMutationRate: Double = options.gaConfig.populationMutationRate
    val geneMutationRate: Double = options.gaConfig.geneMutationRate
    val mutationAmount: Double = options.gaConfig.mutationAmount
    val trialTimes: Int = options.gaConfig.trialTimes
    val population: Population

    var generation: Int
        private set
    private var evolving: Boolean

    private val numberOfSurvival: Int
    private val maxPossibleSnakeLength: Int
    private val multiThreadGames: MultiThreadGames

    init {
        val inputLayerLength = InputLayer.INPUT_LAYER_LENGTH
        val layerShapes = generateLayerShape(
            inputLayerLength,
            *hiddenLayersLength.toIntArray(),
            SnakeBrain.OUTPUT_LAYER_LENGTH
        )

        if (options.providedInfo != null) {
            val (generation, snakeBrains) = options.providedInfo
            if (populationSize != snakeBrains.size) throw Error("Provided snake brains length not equal to population size.")
            this.generation = generation
            population = snakeBrains.map { x ->
                Individual(
                    snakeGame = SnakeGame(
                        options = com.example.snake.game.Options(
                            worldWidth = worldWidth,
                            worldHeight = worldHeight,
                            providedInitialStatus = null,
                        )
                    ),
                    snakeBrain = SnakeBrain(
                        options = com.example.snake.ai.Options(
                            inputLength = x.inputLength,
                            layerShapes = x.layerShapes,
                            hiddenLayerActivationFunction = x.hiddenLayerActivationFunction,
                            providedWeightsAndBiases = ProvidedWeightsAndBiases(x.weights, x.biases)
                        ),
                    ),
                    fitness = 0.0,
                    snakeLength = 1.0,
                    moves = 0.0,
                    survive = true,
                    gameRecord = null
                )
            }.toMutableList()
        } else {
            generation = -1
            population = MutableList(populationSize) {
                Individual(
                    snakeGame = SnakeGame(
                        options = com.example.snake.game.Options(
                            worldWidth = worldWidth,
                            worldHeight = worldHeight,
                            providedInitialStatus = null,
                        )
                    ),
                    snakeBrain = SnakeBrain(
                        com.example.snake.ai.Options(
                            inputLength = inputLayerLength,
                            layerShapes = layerShapes,
                            hiddenLayerActivationFunction = hiddenLayerActivationFunction,
                            providedWeightsAndBiases = null
                        )
                    ),
                    fitness = 0.0,
                    snakeLength = 1.0,
                    moves = 0.0,
                    survive = true,
                    gameRecord = null
                )
            }
        }

        evolving = false
        multiThreadGames = MultiThreadGames()
        numberOfSurvival = (populationSize * surviveRate).toInt()
        if (numberOfSurvival < 2) throw Error("Survival less than 2, please increase survive rate or population size.")
        maxPossibleSnakeLength = worldHeight * worldWidth
    }

    fun exportModel(): GaModelData = GaModelData(
        worldWidth = worldWidth,
        worldHeight = worldHeight,
        hiddenLayersLength = hiddenLayersLength,
        hiddenLayerActivationFunction = hiddenLayerActivationFunction,
        populationSize = populationSize,
        surviveRate = surviveRate,
        populationMutationRate = populationMutationRate,
        geneMutationRate = geneMutationRate,
        mutationAmount = mutationAmount,
        trialTimes = trialTimes,
        generation = generation,
        population = population.map { individual ->
            IndividualPlainObject(
                snakeBrain = individual.snakeBrain.toPlainObject(),
                fitness = individual.fitness,
                snakeLength = individual.snakeLength,
                moves = individual.moves,
                survive = individual.survive,
                gameRecord = individual.gameRecord?.let { SnakeGame.cloneGameRecord(it) }
            )
        }
    )

    suspend fun evolve(): EvolveResult {
        if (evolving) throw Error("Evolve is still running.")

        generation++
        evolving = true
        val startTime = System.currentTimeMillis()

        evaluate()
        select()
        crossover()
        mutate()

        val finalBestPlayer = population.first()

        val bestIndividual = IndividualPlainObject(
            snakeBrain = finalBestPlayer.snakeBrain.toPlainObject(),
            fitness = finalBestPlayer.fitness,
            snakeLength = finalBestPlayer.snakeLength,
            moves = finalBestPlayer.moves,
            survive = finalBestPlayer.survive,
            gameRecord = finalBestPlayer.gameRecord?.let { SnakeGame.cloneGameRecord(it) }
        )

        val overallStats = OverallStats(
            fitness = CalcUtils.statsOfArray(population.map { it.fitness }.toDoubleArray()),
            snakeLength = CalcUtils.statsOfArray(population.map { it.snakeLength }.toDoubleArray()),
            moves = CalcUtils.statsOfArray(population.map { it.moves }.toDoubleArray())
        )

        val timeSpent = System.currentTimeMillis() - startTime
        evolving = false

        return EvolveResult(
            generation,
            bestIndividual,
            timeSpent,
            overallStats
        )
    }

    private suspend fun evaluate() {
        val results = multiThreadGames.playManyGames(
            trialTimes,
            population.map { Pair(it.snakeGame, it.snakeBrain) }
        )

        for (i in population.indices) {
            val individual = population[i]
            val (snakeLengthArr, movesArr, gameRecordArr) = results[i]
            val fitnessArr = Array(trialTimes) { fitness(movesArr[it], snakeLengthArr[it], maxPossibleSnakeLength) }

            individual.snakeLength = CalcUtils.meanOfArray(snakeLengthArr.map { it.toDouble() }.toDoubleArray())
            individual.moves = CalcUtils.meanOfArray(movesArr.map { it.toDouble() }.toDoubleArray())
            individual.fitness = CalcUtils.meanOfArray(fitnessArr.map { it }.toDoubleArray())
            individual.gameRecord =
                gameRecordArr[CalcUtils.indexOfMaxValueInArray(fitnessArr.map { it }.toDoubleArray())]
        }
    }

    private fun select() {
        // sort descending (highest fitness first)
        population.sortByDescending { it.fitness }
        population.forEachIndexed { index, individual -> individual.survive = index < numberOfSurvival }
    }

    private fun crossover() {
        val prefixSum = calculateCumulativeSum(population.map { it.fitness })

        for (i in numberOfSurvival..<populationSize) {
            val parent1 = spinRouletteWheel(population, prefixSum)
            val parent2 = spinRouletteWheel(population, prefixSum)
            population[i].snakeBrain.crossOver(parent1.snakeBrain, parent2.snakeBrain)
        }
    }

    private fun mutate() {
        val best5PercentIndex = (populationSize * 0.05).toInt()

        // e.g. assume populationMutationRate be 0.5, population size be 100, target mutation is 50
        // protect the best 5% from mutating -> 95 * 0.5 / 0.95 = 50
        // the formula is
        // ((p * 0.95) * x) / p = r
        // x = r / 0.95
        // where p is population size, x is population mutation rate, r is target mutation rate
        val mutationRateForRand = populationMutationRate / 0.95

        // prevent the best 5% from mutating
        for (i in best5PercentIndex..<populationSize) {
            if (CalcUtils.randomBool(mutationRateForRand)) {
                population[i].snakeBrain.mutate(geneMutationRate, mutationAmount)
            }
        }
    }
}
